### PR TITLE
detect gnu stat if installed, on osx

### LINKED ISSUE #687: build package on Mac OS shows stat usage error
I build the package on Mac OS and below error will throw:

hchenxa@hchen-mac:~/Documents/hchenxa/src/hchenxa_github/service-catalog$ make build
stat: cannot read file system information for '%c %N': No such file or directory
stat: cannot read file system information for '%c %N': No such file or directory
stat: cannot read file system information for '%c %N': No such file or directory
stat: cannot read file system information for '%c %N': No such file or directory
stat: cannot read file system information for '%c %N': No such file or directory

hchenxa@hchen-mac:~ uname  -a
Darwin hchen-mac.cn.ibm.com 16.5.0 Darwin Kernel Version 16.5.0: Fri Mar  3 16:52:33 PST 2017; root:xnu-3789.51.2~3/RELEASE_X86_64 x86_64 i386 MacBookPro11,5 Darwin

### PR DESCRIPTION
**Pull request by MHBauer**
Fixes #687

### CODE REVIEW

**Comment by duglin** [comment]
you crazy native mac users :-)
LGTM

**Comment by krancour** [comment]
This introduces a pretty specific and inflexible requirement for how people need their local systems set up in order to be successful. I am 100% against that. I believe their are better ways of determining which implementation of a given tool (e.g. `stat`) is available. Let me go looking for it... I think it's already used in at least one place.

**Comment by krancour** [comment]
@MHBauer see here: https://github.com/kubernetes-incubator/service-catalog/pull/690#issuecomment-293600459

**Comment by MHBauer** [comment]
ah, okay, I'll take a look at that.

File: Makefile [hunk 1]
```diff
@@ -33,8 +33,10 @@ SRC_DIRS       = $(shell sh -c "find $(TOP_SRC_DIRS) -name \\*.go \
 TEST_DIRS     ?= $(shell sh -c "find $(TOP_SRC_DIRS) -name \\*_test.go \
                    -exec dirname {} \\; | sort | uniq")
 VERSION       ?= $(shell git describe --always --abbrev=7 --dirty)
+# on osx, check for gnu-stat installed as gstat, and if not use the
+# built-in bsd-stat. Otherwise, assume building on linux with gnu-stat.
 ifeq ($(shell uname -s),Darwin)
-STAT           = stat -f '%c %N'
+STAT           = $(if command -v gstat, stat -c '%Y %n', stat -f '%c %N')
```
Comment (MHBauer) [inline Makefile]: oops. that should have been gstat for the first stat invocation string.

**Comment by MHBauer** [comment]
re-review.

File: Makefile [hunk 2]
```diff
@@ -34,10 +34,17 @@ TEST_DIRS     ?= $(shell sh -c "find $(TOP_SRC_DIRS) -name \\*_test.go \
...
+# output.  if stdout is blank, bsd-stat because it returned an error
+# to stderr. Otherwise use gnu-stat args.
+ifeq ($(shell stat -c"%U" /dev/null 2> /dev/null),)
 STAT           = stat -f '%c %N'
 else
 STAT           = stat -c '%Y %n'
 endif
+else
+STAT           = stat -c '%Y %n'
+endif
```
Comment (krancour) [inline Makefile]: I would recommend simplifying all this. The bit about being Darwin or not isn't relevant in comparison to simply knowing whether we're dealing with gnu or bsd stat. i.e. `ifeq ($(shell stat -c"%U" /dev/null 2> /dev/null),)`, should always happen be evaluated regardless of OS.

Comment (MHBauer) [inline Makefile]: ah, make sense.

**Comment by MHBauer** [comment]
@krancour @duglin done, built. Checked locally with both gnu stat and bsd stat.

File: Makefile [hunk 3]
```diff
@@ -33,11 +33,16 @@ SRC_DIRS       = $(shell sh -c "find $(TOP_SRC_DIRS) -name \\*.go \
 TEST_DIRS     ?= $(shell sh -c "find $(TOP_SRC_DIRS) -name \\*_test.go \
                    -exec dirname {} \\; | sort | uniq")
 VERSION       ?= $(shell git describe --always --abbrev=7 --dirty)
-ifeq ($(shell uname -s),Darwin)
+
+# run stat against /dev/null and check if it has any stdout output.
+# if stdout is blank, bsd-stat because stat returned an error to
```
Comment (duglin) [inline Makefile]: missing a "use"  before bsd-stat?

Comment (MHBauer) [inline Makefile]: if you want. I can capitalize some things too.

**Comment by MHBauer** [comment]
@duglin rewrote

**Comment by krancour** [comment]
Veto rescinded. :) LGTM

**Comment by duglin** [comment]
still LGTM

**Pull request merged**